### PR TITLE
Zero credential data before free in NLA and NTLM context

### LINKED ISSUE #12531: Credential data not zeroed before free in NLA UTF-16 conversion and NTLM context
## Summary

Temporary UTF-16 copies of credentials (password, username, domain) in `libfreerdp/core/nla.c` are `free()`d without being zeroed first. Additionally, `ntlm_ContextFree()` frees the NTLM context struct without zeroing the embedded `NtlmHash[16]` and `NtlmV2Hash[16]` arrays. This leaves plaintext credential data in freed heap memory, recoverable via memory analysis tools.

## Context

FreeRDP already has thorough credential zeroing in most code paths:

- `sspi_FreeAuthIdentity()` uses `zfree()` which `memset`s before `free` ([sspi.c:963-968](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/winpr/libwinpr/sspi/sspi.c#L963))
- `sspi_CredentialsFree()` explicitly `memset`s User/Domain/Password before freeing ([sspi_winpr.c:212-220](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/winpr/libwinpr/sspi/sspi_winpr.c#L212))
- `sspi_SecBufferFree()` zeroes `pvBuffer` before freeing ([sspi_winpr.c:243-244](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/winpr/libwinpr/sspi/sspi_winpr.c#L238))
- `free_string()` in settings zeroes when `cleanup=TRUE` ([settings_getters.c:17-23](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/libfreerdp/common/settings_getters.c#L15))

Previous issues (#2904, #3639) addressed credential exposure via `/proc/pid/cmdline`. The command-line password is now properly scrubbed. However, the in-memory credential copies created during NLA/CredSSP authentication were not given the same treatment.

### 1. `libfreerdp/core/nla.c` — `nla_encode_ts_credentials()`

Four calls to `freerdp_settings_get_string_as_utf16()` allocate new UTF-16 buffers for credentials, which are freed with `free()` without zeroing:

| Line | Setting | Issue |
|------|---------|-------|
| [1662–1665](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/libfreerdp/core/nla.c#L1662) | `FreeRDP_Password` | `free(octet_string.data)` without `memset` |
| [1684–1691](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/libfreerdp/core/nla.c#L1684) | `FreeRDP_Password` (alternate path) | same |
| [1705–1708](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/libfreerdp/core/nla.c#L1705) | `FreeRDP_Username` | same |
| [1717–1720](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/libfreerdp/core/nla.c#L1717) | `FreeRDP_Domain` | same |

### 2. `winpr/libwinpr/sspi/NTLM/ntlm.c` — `ntlm_ContextFree()`

[Line 364](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/winpr/libwinpr/sspi/NTLM/ntlm.c#L364): `free(context)` without zeroing the embedded `NtlmHash[16]` and `NtlmV2Hash[16]` byte arrays (struct fields at [ntlm.h:228-229](https://github.com/FreeRDP/FreeRDP/blob/3.24.1/winpr/libwinpr/sspi/NTLM/ntlm.h#L228)).

## How to reproduce

1. Build any FreeRDP client (e.g. `xfreerdp`) in Debug mode
2. Launch with `MallocStackLogging=1` (macOS) or equivalent heap tracing
3. Connect to an RDP server with NLA authentication
4. Disconnect
5. Scan the process memory for the password string — copies remain in freed heap pages

We confirmed this using an automated test harness that scans writable process memory regions via Mach VM APIs and traces each match to its allocation stack via `malloc_history`. After a connect/disconnect cycle, 5 password copies from `mac_client_thread` → `freerdp_connect` remain in freed-but-not-zeroed heap memory.

## Suggested fix

Replace bare `free()` calls with `memset` + `free` for credential-containing buffers, consistent with the existing pattern used throughout FreeRDP's SSPI layer.

In `nla.c`, replace:
```c
free(octet_string.data);
```
with:
```c
if (octet_string.data)
    memset(octet_string.data, 0, octet_string.len);
free(octet_string.data);
```

In `ntlm.c` `ntlm_ContextFree()`, add before `free(context)`:
```c
memset(context->NtlmHash, 0, sizeof(context->NtlmHash));
memset(context->NtlmV2Hash, 0, sizeof(context->NtlmV2Hash));
```

This is a minimal, targeted change (~10 lines) that aligns these code paths with the zeroing discipline already established in `sspi_FreeAuthIdentity`, `sspi_CredentialsFree`, and `sspi_SecBufferFree`.

## Impact

security hardening only. No functional change. Prevents credential recovery from process memory dumps after RDP session disconnect.

### PR DESCRIPTION
**Pull request by Kotivskyi**
## Summary

- Zero temporary UTF-16 credential copies in `nla_encode_ts_smartcard_credentials()` before `free()`
- Zero NTLM hash and session key material in `ntlm_ContextFree()` before `free(context)`
- Add `TestCredentialZeroing` test verifying `sspi_FreeAuthIdentity` and `sspi_SecBufferFree` properly zero fields

## Problem

`freerdp_settings_get_string_as_utf16()` allocates new UTF-16 buffers for credentials in `nla.c`. These were `free()`d without `memset`, leaving plaintext password/username/domain data in freed heap memory — recoverable via memory analysis tools (`malloc_history`, core dumps, etc.).

Similarly, `ntlm_ContextFree()` freed the NTLM context without zeroing embedded `NtlmHash[16]`, `NtlmV2Hash[16]`, and session key arrays.

FreeRDP already has thorough credential zeroing elsewhere:
- `sspi_FreeAuthIdentity()` → `zfree()` (memset + free)
- `sspi_CredentialsFree()` → explicit memset before free
- `sspi_SecBufferFree()` → memset before free
- `free_string()` in settings → memset when `cleanup=TRUE`

These specific code paths were simply missed.

## Changes

**`libfreerdp/core/nla.c`** — 4 `free(octet_string.data)` calls now preceded by `memset`:
- Line 1665: `FreeRDP_Password` (UTF-16, smartcard pin)
- Line 1691: `FreeRDP_Password` (UTF-16, CSP data loop)
- Line 1708: `FreeRDP_Username` (UTF-16, userHint)
- Line 1720: `FreeRDP_Domain` (UTF-16, domainHint)

**`winpr/libwinpr/sspi/NTLM/ntlm.c`** — `ntlm_ContextFree()` now zeroes:
- `NtlmHash[16]`, `NtlmV2Hash[16]`
- `SessionBaseKey`, `KeyExchangeKey`, `RandomSessionKey`, `ExportedSessionKey`, `EncryptedRandomSessionKey`
- `NtProofString`

**`winpr/libwinpr/sspi/test/TestCredentialZeroing.c`** — New test validating:
- `sspi_FreeAuthIdentity` zeroes and NULLs all identity fields
- `sspi_SecBufferFree` zeroes and NULLs buffer
- NULL/empty input handling doesn't crash

## Test plan

- [x] `TestCredentialZeroing` passes
- [x] `TestNTLM` passes (no regression)
- [x] Verified with macOS `MallocStackLogging` + `malloc_history` tracing

Fixes #12531

### CODE REVIEW

**Comment by Kotivskyi** [comment]
## Test Results

Full test suite run on macOS (arm64, FreeRDP 3.24.1 tag):

```
99% tests passed, 1 tests failed out of 155

Total Test time (real) =   9.00 sec

The following tests FAILED:
    105 - TestThreadCreateProcess (Failed)
```

**154/155 pass.** The single failure (`TestThreadCreateProcess`) is a pre-existing macOS platform issue — it fails identically on the unmodified 3.24.1 tag, unrelated to this PR.

### SSPI tests (all pass)

| Test | Status |
|------|--------|
| `TestQuerySecurityPackageInfo` | PASS |
| `TestEnumerateSecurityPackages` | PASS |
| `TestInitializeSecurityContext` | PASS |
| `TestAcquireCredentialsHandle` | PASS |
| `TestCredSSP` | PASS |
| `TestNTLM` | PASS |
| `TestCredentialZeroing` (new) | PASS |

### Verification methodology

The issue was originally found using an automated test harness that:
1. Launches an RDP client with `MallocStackLogging=1`
2. Connects to an RDP server (NLA authentication)
3. Disconnects
4. Scans writable process memory for plaintext password using Mach VM APIs
5. Traces each match to its allocation stack via `malloc_history`

Before this patch: 5 password copies from `mac_client_thread` → `freerdp_connect` remained in freed heap memory after disconnect, traced to `nla_encode_ts_smartcard_credentials()` UTF-16 conversions and NTLM context key material.

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by Kotivskyi** [comment]
Fixed CMake-format error, sorry for the force push

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/nla.c [hunk 1]
```diff
@@ -1662,6 +1662,8 @@ static BOOL nla_encode_ts_smartcard_credentials(rdpNla* nla, WinPrAsn1Encoder* e
 	    (BYTE*)freerdp_settings_get_string_as_utf16(settings, FreeRDP_Password, &ss);
 	octet_string.len = ss * sizeof(WCHAR);
 	BOOL res = WinPrAsn1EncContextualOctetString(enc, 0, &octet_string) > 0;
+	if (octet_string.data)
```
Comment (akallabeth) [inline libfreerdp/core/nla.c]: There is already `WinPrAsn1FreeOctetString`, maybe better use that here and fix the missing zero there?

File: libfreerdp/core/nla.c [hunk 2]
```diff
@@ -1688,6 +1690,8 @@ static BOOL nla_encode_ts_smartcard_credentials(rdpNla* nla, WinPrAsn1Encoder* e
 		{
 			const BOOL res2 =
 			    WinPrAsn1EncContextualOctetString(enc, cspData_fields[i].tag, &octet_string) > 0;
+			if (octet_string.data)
+				memset(octet_string.data, 0, octet_string.len);
```
Comment (akallabeth) [inline libfreerdp/core/nla.c]: see above

File: libfreerdp/core/nla.c [hunk 3]
```diff
@@ -1705,6 +1709,8 @@ static BOOL nla_encode_ts_smartcard_credentials(rdpNla* nla, WinPrAsn1Encoder* e
 		    (BYTE*)freerdp_settings_get_string_as_utf16(settings, FreeRDP_Username, &ss);
 		octet_string.len = ss * sizeof(WCHAR);
 		res = WinPrAsn1EncContextualOctetString(enc, 2, &octet_string) > 0;
+		if (octet_string.data)
```
Comment (akallabeth) [inline libfreerdp/core/nla.c]: see above

File: libfreerdp/core/nla.c [hunk 4]
```diff
@@ -1717,6 +1723,8 @@ static BOOL nla_encode_ts_smartcard_credentials(rdpNla* nla, WinPrAsn1Encoder* e
 		    (BYTE*)freerdp_settings_get_string_as_utf16(settings, FreeRDP_Domain, &ss);
 		octet_string.len = ss * sizeof(WCHAR);
 		res = WinPrAsn1EncContextualOctetString(enc, 3, &octet_string) > 0;
+		if (octet_string.data)
```
Comment (akallabeth) [inline libfreerdp/core/nla.c]: see above

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	identity.DomainLength = (UINT32)strlen(testDomain);
+	identity.Domain = (UINT16*)strdup(testDomain);
+	if (!identity.Domain)
+	{
+		free(identity.User);
+		return FALSE;
+	}
+
+	identity.PasswordLength = (UINT32)strlen(testPassword);
+	identity.Password = (UINT16*)strdup(testPassword);
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `ConvertUtf8ToWCharAlloc` maybe?

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	if (!identity.Password)
+	{
+		free(identity.User);
+		free(identity.Domain);
+		return FALSE;
+	}
+
+	sspi_FreeAuthIdentity(&identity);
+
+	if (identity.User != NULL)
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: we added `C23` support, replace `NULL` with `nullptr` (we provide a fallback definition for `C11`)

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	{
+		printf("FAIL: identity.User not NULL after FreeAuthIdentity\n");
+		return FALSE;
+	}
+	if (identity.Domain != NULL)
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: see above

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	{
+		printf("FAIL: identity.Domain not NULL after FreeAuthIdentity\n");
+		return FALSE;
+	}
+	if (identity.Password != NULL)
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: see above

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	return TRUE;
+}
+
+/* Test that sspi_SecBufferFree zeroes and NULLs the buffer */
+static BOOL test_SecBufferFree_zeroes_buffer(void)
+{
+	const char* testData = "sensitive-credential-data";
+	SecBuffer buffer = { 0 };
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: using `WINPR_C_ARRAY_INIT` define for this

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+	buffer.cbBuffer = (ULONG)strlen(testData);
+	buffer.pvBuffer = strdup(testData);
+	if (!buffer.pvBuffer)
+		return FALSE;
+
+	sspi_SecBufferFree(&buffer);
+
+	if (buffer.pvBuffer != NULL)
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `nullptr`

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	return TRUE;
+}
+
+/* Test NULL and empty input handling */
+static BOOL test_null_handling(void)
+{
+	/* NULL should not crash */
+	sspi_FreeAuthIdentity(NULL);
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `nullptr`

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	sspi_SecBufferFree(NULL);
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `nullptr`

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+	/* Empty identity should not crash */
+	SEC_WINNT_AUTH_IDENTITY empty = { 0 };
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `WINPR_C_ARRAY_INIT`

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+	sspi_FreeAuthIdentity(&empty);
+
+	/* Empty buffer should not crash */
+	SecBuffer emptyBuf = { 0 };
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `WINPR_C_ARRAY_INIT`

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+ */
+
+/* Test that sspi_FreeAuthIdentity zeroes and NULLs all fields */
+static BOOL test_FreeAuthIdentity_zeroes_fields(void)
+{
+	const char* testPassword = "S3cretP@ssw0rd!";
+	const char* testUser = "testuser";
+	const char* testDomain = "TESTDOMAIN";
+
+	SEC_WINNT_AUTH_IDENTITY identity = { 0 };
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: `WINPR_C_ARRAY_INIT`

File: libfreerdp/core/nla.c (hunk 1 above)
Comment (Kotivskyi) [reply libfreerdp/core/nla.c]: Agree. Fixed

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c (hunk 6 above)
Comment (Kotivskyi) [reply winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: fixed

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c (hunk 9 above)
Comment (Kotivskyi) [reply winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: fixed

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c (hunk 14 above)
Comment (Kotivskyi) [reply winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: fixed

File: winpr/libwinpr/sspi/test/TestCredentialZeroing.c (hunk 5 above)
Comment (Kotivskyi) [reply winpr/libwinpr/sspi/test/TestCredentialZeroing.c]: Agree. Fixed

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@Kotivskyi thank you for your contribution!
